Backup FileBrowser: isolate state per backup rewindId

Initsogar opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Resolves DOTDASH-533

## Proposed Changes

* Replace useState with `Map<rewindId, FileBrowserState>` for per-backup state isolation
* Update all `FileBrowserStateActions` interface methods to require a `rewindId` parameter
* Update all consumer components to pass `rewindId` to file browser state methods
  * Staging site sync (V1 and V2)
  * Backup's page in Jetpack Cloud
  * Backup's page in Hosting Dashboard

| Before | After |
|---|---|
| <video src="https://github.com/user-attachments/assets/b3232dc3-07ae-43b0-9c8e-87bc64b7c60c" /> | <video src="https://github.com/user-attachments/assets/d30b7ff7-aa71-42bf-8dba-3425d80d367b" /> |

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

The file browser state was persisting across different backups, causing file selections to bleed between them. Selecting files in backup A would also appear as selected when viewing backup B. This is an specific issue in the Backups' HD page, it doesn't impact other consumers (staging sync modal, or Jetpack Cloud backup contents page).

This isolation ensures each backup gets its own independent file browser state, eliminating cross-contamination and providing a clean user experience when switching between different backup points.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Test state Isolation:
  - Navigate to a backup (backup A) and select some files/folders
  - Switch to a different backup (backup B)
  - Verify selections do NOT persist (backup B should have clean state)
  - Try downloading some files, verify the files you select are the ones you got in the download.
  - Try restoring some files, verify the files you select are the ones you see in the restore confirmation page.
2. Ensure staging sync modal keeps working as expected
3. Ensure backup page in Jetpack Cloud keeps working as expected

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

Initsogar committed:
Isolate file browser state per backup rewindId

- Replace useState with Map<rewindId, state> for true isolation
- Update all FileBrowserStateActions to require rewindId parameter
- Add FileBrowser key prop to force remount on backup changes
- Prevents file selections from bleeding between different backups

Initsogar commented on client/dashboard/sites/staging-site-sync-modal/index.tsx:
```diff
@@ -206,16 +206,32 @@ function StagingSiteSyncModalInner( {
 		onClose();
 	}, [ onClose ] );
 
+	// Determine latest successful backup for rewindId/display
+	const activityQuery = useRewindableActivityLogQuery(
+		querySiteId,
+		{
+			name: SUCCESSFUL_BACKUP_ACTIVITIES,
+			aggregate: false,
+			number: 1,
+			sortOrder: 'desc',
+		},
+		{}
+	) as { data: BackupActivity[] | undefined; isLoading: boolean };
+	const { data: backupData, isLoading: isLoadingBackupAttempt } = activityQuery;
+
+	const lastKnownBackupAttempt: BackupActivity | undefined = backupData?.[ 0 ];
```
This is moved from line 272, so we have access to rewindId early for the file browser operations.

Initsogar committed:
Fix granular restore on Jetpack Cloud

Initsogar committed:
Merge branch 'trunk' into update/hd-backup-file-browser-rewind-id

katinthehatsite reviewed (approved):
I only tested this for the old and new staging sync modal and did not find any regresssions from that perspective.

katinthehatsite commented on client/dashboard/sites/backup-restore/granular-form.tsx:
```diff
@@ -26,8 +26,8 @@ function SiteBackupGranularRestoreForm( {
 	const { fileBrowserState } = useFileBrowserContext();
-	const browserCheckList = fileBrowserState.getCheckList();
-	const browserSelectedList = fileBrowserState.getSelectedList();
+	const browserCheckList = fileBrowserState.getCheckList( Number( rewindId ) );
```
Do you think we need to do this type conversion here or can it be skipped? The rewindId is already typed as `number` in the component props, and the methods expect `number` parameters.

Initsogar replied:
I think you mean [BackupContentsPage](https://github.com/Automattic/wp-calypso/pull/106035/commits/1df20f14405b3c043c89fa774ed37c2df91321a3#diff-decf10530fe9d64786be541942b4060e6265bf1418b03f1785d58a11ef912151L30)? It looks like we were misrepresenting the types, since we expect a number but were sending any from the URL params. I addressed this in https://github.com/Automattic/wp-calypso/pull/106035/commits/2cf295f57fafa80cb764686001c33f45b6fbc472.

Initsogar committed:
Refactor rewindId handling in BackupContentsPage

- Convert rewindId to a number in the controller for consistency.
- Ensure file browser state retrieval uses rewindId directly without conversion.

Initsogar committed:
Merge remote-tracking branch 'origin/trunk' into update/hd-backup-file-browser-rewind-id

Initsogar committed:
Merge remote-tracking branch 'refs/remotes/origin/update/hd-backup-file-browser-rewind-id' into update/hd-backup-file-browser-rewind-id

myhro reviewed (approved):
Worked exactly like the shared screencasts. LGTM. ✅

Initsogar committed:
Merge branch 'trunk' into update/hd-backup-file-browser-rewind-id
